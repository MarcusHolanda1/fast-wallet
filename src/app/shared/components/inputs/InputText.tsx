import { theme } from '@shared/theme/theme';
import React from 'react';
import {
  View,
  TextInput,
  StyleSheet,
  TextInputProps,
  Text
} from 'react-native';
import MaskInput, { MaskInputProps } from 'react-native-mask-input';

interface InputTextProps extends Omit<TextInputProps, 'onChangeText'> {
  icon?: React.ReactNode;
  containerStyle?: object;
  label: string;
  mask?: MaskInputProps['mask'];
  onChangeText?: (masked: string, unmasked: string) => void;
  errorText?: string;
  prefix?: React.ReactNode;
}

const InputText: React.FC<InputTextProps> = ({
  icon,
  containerStyle,
  style,
  label,
  mask,
  errorText,
  prefix,
  onChangeText,
  ...props
}) => {
  return (
    <View>
      <Text style={styles.label}>{label}</Text>
      <View style={[styles.container, containerStyle]}>
        {prefix}
        {mask ? (
          <MaskInput
            style={[styles.input, style]}
            mask={mask}
            onChangeText={onChangeText}
            {...props}
          />
        ) : (
          <TextInput
            style={[styles.input, style]}
            onChangeText={(text) => onChangeText?.(text, text)}
            {...props}
          />
        )}
        {icon && <View style={styles.iconContainer}>{icon}</View>}
      </View>
      <Text style={styles.errorText}>{errorText}</Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: 45,
    borderRadius: 6,
    flexDirection: 'row',
    alignItems: 'center',
    marginVertical: 8,
    borderWidth: 1,
    borderColor: theme.colors.base.greyLight,
    backgroundColor: theme.colors.base.white
  },
  input: {
    ...theme.typography.p,
    color: theme.colors.text.black,
    backgroundColor: 'transparent',
    paddingHorizontal: 16,
    flex: 1
  },
  iconContainer: {
    marginRight: 12,
    justifyContent: 'center',
    alignItems: 'center'
  },
  label: {
    color: theme.colors.text.grey,
    fontSize: theme.typography.pSmall.fontSize
  },
  errorText: {
    color: theme.colors.alert.red,
    fontSize: theme.typography.pSmall.fontSize,
    marginBottom: 4
  }
});

export default InputText;
