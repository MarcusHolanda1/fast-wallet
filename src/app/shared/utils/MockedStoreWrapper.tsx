import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';

const createTestStore = () => {
  return configureStore({
    reducer: {
      test: (state: Record<string, unknown> = {}) => state
    }
  });
};

export const TestWrapper = ({ children }: { children: React.ReactNode }) => {
  const store = createTestStore();
  return <Provider store={store}>{children}</Provider>;
};
