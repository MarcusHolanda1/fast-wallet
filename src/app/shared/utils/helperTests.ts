import { fireEvent, RenderResult } from '@testing-library/react-native';

export const fillAndBlur = (
  sut: RenderResult,
  testId: string,
  value: string
) => {
  fireEvent.changeText(sut.getByTestId(testId), value);
  fireEvent(sut.getByTestId(testId), 'blur');
};

export const expectFormattedCardNumber = (cardNumber: string) => {
  return '•••• •••• •••• ' + cardNumber.slice(-4);
};
