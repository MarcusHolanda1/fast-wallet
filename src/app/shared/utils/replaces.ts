export const replaceWithTextOnly = (text: string) => {
  return text.replace(/[^a-zA-ZÃ€-Ã¿\s]/g, '');
};

export const formatCardNumber = (cardNumber: string) => {
  return cardNumber.slice(-4);
};
