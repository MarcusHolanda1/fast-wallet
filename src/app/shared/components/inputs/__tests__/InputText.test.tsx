import { fireEvent, render, RenderResult } from '@testing-library/react-native';
import { TestWrapper } from '@app/shared/utils/MockedStoreWrapper';
import { Text } from 'react-native';

import InputText from '../InputText';

const mockOnChangeText = jest.fn();

const makeSut = (props = {}) => {
  const defaultProps = {
    label: 'Test Label',
    onChangeText: mockOnChangeText,
    ...props
  };

  return render(
    <TestWrapper>
      <InputText {...defaultProps} />
    </TestWrapper>
  );
};

describe('InputText Component', () => {
  let sut: RenderResult;

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Rendering', () => {
    test('should render label correctly', () => {
      sut = makeSut({ label: 'Email Address' });
      expect(sut.getByText('Email Address')).toBeTruthy();
    });

    test('should render with placeholder', () => {
      sut = makeSut({ placeholder: 'Enter your email' });
      expect(sut.getByPlaceholderText('Enter your email')).toBeTruthy();
    });

    test('should render error text when provided', () => {
      sut = makeSut({ errorText: 'This field is required' });
      expect(sut.getByText('This field is required')).toBeTruthy();
    });

    test('should not render error text when not provided', () => {
      sut = makeSut();
      expect(sut.queryByTestId('error-text')).toBeNull();
    });

    test('should render icon when provided', () => {
      const mockIcon = <Text testID="test-icon">Icon</Text>;
      sut = makeSut({ icon: mockIcon });
      expect(sut.getByTestId('test-icon')).toBeOnTheScreen();
    });

    test('should render prefix when provided', () => {
      const mockPrefix = <Text testID="test-prefix">$</Text>;
      sut = makeSut({ prefix: mockPrefix });
      expect(sut.getByTestId('test-prefix')).toBeOnTheScreen();
    });
  });

  describe('Text Input Functionality', () => {
    test('should call onChangeText when text is entered (without mask)', () => {
      sut = makeSut();

      fireEvent.changeText(sut.getByDisplayValue(''), 'test text');

      expect(mockOnChangeText).toHaveBeenCalledWith('test text', 'test text');
    });

    test('should handle empty onChangeText prop', () => {
      sut = makeSut({ onChangeText: undefined });
      expect(() =>
        fireEvent.changeText(sut.getByDisplayValue(''), 'test')
      ).not.toThrow();
    });

    test('should render with initial value', () => {
      sut = makeSut({ value: 'Initial Value' });
      expect(sut.getByDisplayValue('Initial Value')).toBeTruthy();
    });

    test('should handle focus and blur events', () => {
      const mockOnFocus = jest.fn();
      const mockOnBlur = jest.fn();

      sut = makeSut({ onFocus: mockOnFocus, onBlur: mockOnBlur });

      fireEvent(sut.getByDisplayValue(''), 'focus');
      expect(mockOnFocus).toHaveBeenCalled();

      fireEvent(sut.getByDisplayValue(''), 'blur');
      expect(mockOnBlur).toHaveBeenCalled();
    });
  });

  describe('Masked Input Functionality', () => {
    test('should render MaskInput when mask is provided', () => {
      const mask = [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/];
      sut = makeSut({ mask });

      fireEvent.changeText(sut.getByDisplayValue(''), '12/34/5678');

      expect(mockOnChangeText).toHaveBeenCalled();
    });

    test('should handle masked input changes', () => {
      const mask = [/\d/, /\d/, '/', /\d/, /\d/];
      sut = makeSut({ mask });

      fireEvent.changeText(sut.getByDisplayValue(''), '1234');

      expect(mockOnChangeText).toHaveBeenCalled();
    });
  });

  describe('Styling and Props', () => {
    test('should apply custom container style', () => {
      const customStyle = { backgroundColor: 'red' };
      sut = makeSut({ containerStyle: customStyle });

      expect(sut.getByText('Test Label')).toBeTruthy();
    });

    test('should apply custom input style', () => {
      const customStyle = { fontSize: 20 };
      sut = makeSut({ style: customStyle });

      expect(sut.getByText('Test Label')).toBeTruthy();
    });

    test('should pass through TextInput props', () => {
      sut = makeSut({
        autoCapitalize: 'none',
        keyboardType: 'email-address',
        secureTextEntry: true
      });

      expect(sut.getByText('Test Label')).toBeTruthy();
    });

    test('should handle multiline prop', () => {
      sut = makeSut({ multiline: true });
      expect(sut.getByText('Test Label')).toBeTruthy();
    });

    test('should handle editable prop', () => {
      sut = makeSut({ editable: false });
      expect(sut.getByText('Test Label')).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    test('should handle null/undefined values gracefully', () => {
      sut = makeSut({
        value: null,
        errorText: undefined,
        icon: null,
        prefix: undefined
      });

      expect(sut.getByText('Test Label')).toBeTruthy();
    });

    test('should handle empty string values', () => {
      sut = makeSut({
        value: '',
        errorText: '',
        label: ''
      });

      expect(sut.getByDisplayValue('')).toBeTruthy();
    });

    test('should render both prefix and icon together', () => {
      const mockPrefix = <Text testID="test-prefix">$</Text>;
      const mockIcon = <Text testID="test-icon">Icon</Text>;

      sut = makeSut({ prefix: mockPrefix, icon: mockIcon });

      expect(sut.getByTestId('test-prefix')).toBeOnTheScreen();
      expect(sut.getByTestId('test-icon')).toBeOnTheScreen();
    });
  });

  test('should match snapshot', () => {
    sut = makeSut({
      label: 'Snapshot Label',
      placeholder: 'Snapshot Placeholder',
      errorText: 'Snapshot Error',
      icon: <Text>Icon</Text>,
      prefix: <Text>$</Text>,
      style: { fontSize: 18 },
      containerStyle: { padding: 10 },
      value: 'Snapshot Value',
      mask: [/\d/, /\d/, '/', /\d/, /\d/, '/', /\d/, /\d/, /\d/, /\d/]
    });
    expect(sut.toJSON()).toMatchSnapshot();
  });
});
