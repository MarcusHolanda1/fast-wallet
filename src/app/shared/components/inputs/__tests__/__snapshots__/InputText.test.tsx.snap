// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputText Component should match snapshot 1`] = `
<View>
  <Text
    style={
      {
        "color": "#BBBBBB",
        "fontSize": 14,
      }
    }
  >
    Snapshot Label
  </Text>
  <View
    style={
      [
        {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderColor": "#EEEEEE",
          "borderRadius": 6,
          "borderWidth": 1,
          "flexDirection": "row",
          "height": 45,
          "marginVertical": 8,
        },
        {
          "padding": 10,
        },
      ]
    }
  >
    <Text>
      $
    </Text>
    <TextInput
      onChangeText={[Function]}
      placeholder="Snapshot Placeholder"
      style={
        [
          {
            "backgroundColor": "transparent",
            "color": "#111111",
            "flex": 1,
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 18,
            "paddingHorizontal": 16,
          },
          {
            "fontSize": 18,
          },
        ]
      }
      value=""
    />
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "marginRight": 12,
        }
      }
    >
      <Text>
        Icon
      </Text>
    </View>
  </View>
  <Text
    style={
      {
        "color": "#FF0000",
        "fontSize": 14,
        "marginBottom": 4,
      }
    }
    testID="error-text"
  >
    Snapshot Error
  </Text>
</View>
`;
